snapcraft/parts/update_metadata.py

```diff
@@ -0,0 +1,163 @@
+from snapcraft.projects import MANDATORY_ADOPTABLE_FIELDS, Project
+
+
+def update_project_metadata(
+    project: Project,
+    *,
+    project_vars: Dict[str, str],
+    metadata_list: List[ExtractedMetadata],
+    assets_dir: Path,
+    prime_dir: Path,
+) -> None:
+    """Set project fields using corresponding adopted entries.
+
+    Fields are validated on assignment by pydantic.
+
+    :param project: The project to update.
+    :param project_vars: The variables updated during lifecycle execution.
+    :param metadata_list: List containing parsed information from metadata files.
+
+    :raises SnapcraftError: If project update failed.
+    """
+    _update_project_variables(project, project_vars)
+
+    for metadata in metadata_list:
+        # Data specified in the project yaml has precedence over extracted data
+        if metadata.title and not project.title:
+            project.title = metadata.title
+
+        if metadata.summary and not project.summary:
+            project.summary = metadata.summary
+
+        if metadata.description and not project.description:
+            project.description = metadata.description
+
+        if metadata.version and not project.version:
+            project.version = metadata.version
+
+        if metadata.grade and not project.grade:
+            project.grade = metadata.grade  # type: ignore
+
+        if not project.icon:
+            _update_project_icon(
+                project, metadata=metadata, assets_dir=assets_dir, prime_dir=prime_dir
+            )
+
+        _update_project_app_desktop_file(
+            project, metadata=metadata, assets_dir=assets_dir, prime_dir=prime_dir
+        )
+
+    # Fields that must not end empty
+    for field in MANDATORY_ADOPTABLE_FIELDS:
+        if not getattr(project, field):
+            raise errors.SnapcraftError(
+                f"Field {field!r} was not adopted from metadata"
+            )
+
+
+def _update_project_variables(project: Project, project_vars: Dict[str, str]):
+    """Update project fields with values set during lifecycle processing."""
+    try:
+        if project_vars["version"]:
+            project.version = project_vars["version"]
+        if project_vars["grade"]:
+            project.grade = project_vars["grade"]  # type: ignore
+    except pydantic.ValidationError as err:
+        _raise_formatted_validation_error(err)
+        raise errors.SnapcraftError(f"error setting variable: {err}")
+
+
+def _update_project_icon(
+    project: Project, *, metadata: ExtractedMetadata, assets_dir: Path, prime_dir: Path
+) -> None:
+    """Look for icons files and update project.
+
+    Existing icon in snap/gui/icon.{png,svg} has precedence over extracted data
+    """
+    for icon_file in (f"{assets_dir}/gui/icon.png", f"{assets_dir}/gui/icon.svg"):
+        if Path(icon_file).is_file():
+            break
+    else:
+        if metadata.icon and Path(prime_dir, metadata.icon).is_file():
+            project.icon = metadata.icon
```

Comment:
Somewhere above `VALID_ICONS_EXTENSIONS: Final[List[str]] = ["png", "svg"]`
then,
```suggestion
    icon_files = (Path(f"{assets_dir}/gui/icon.{ext}") for ext in VALID_ICON_EXTENSIONS)
    existing_icon_files = [p for p in icon_files if p.is_file()]
    if metadata.icon and Path(prime_dir, metadata.icon).is_file():
        existing_icon_files.append(metadata.icon)
    
    with contextlib.suppress(IndexError):
        project.icon = existing_icon_files[0]
```

This might simplify the logic (one place) to decide how to copy the icon over.

Reply:
I like the idea of defining valid icon extensions separately. I'll refactor the code a bit to use the definition, but I'd avoid testing for the existence of all files when just testing the first one could be enough if it satisfies the exit condition. The icon copying code probably won't be affected by this implementation.